### PR TITLE
[#1563] Fixed Line Header Code Minings at Pos 0,0

### PR DESCRIPTION
**Pull request by cdietrich**
[#1563] Fixed Line Header Code Minings at Pos 0,0
Fixes: #1563

Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

**Comment by miklossy** [comment]
Thanks for the fix @cdietrich . I tested your improvement and seems to work. Nice! 

However, I think it would worth extending the [DomainmodelCodeMiningProvider](https://github.com/eclipse/xtext-eclipse/blob/master/org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src/org/eclipse/xtext/example/domainmodel/ui/codemining/DomainmodelCodeMiningProvider.java) and the [CodeMiningTest](https://github.com/eclipse/xtext-eclipse/blob/master/org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui.tests/src/org/eclipse/xtext/example/domainmodel/ui/tests/CodeMiningTest.xtend) by corresponding test cases as described in the issue report.

**Comment by cdietrich** [comment]
please do so in a separate PR

**Comment by miklossy** [comment]
I have tested your fix with the Xtext Domainmodel and the Simple Arithmetics examples as well as with a customer code base. All tests are passing!

**Review by miklossy — APPROVED** [review]

**Comment by miklossy** [comment]
> 
> 
> please do so in a separate PR

https://github.com/eclipse/xtext-eclipse/pull/1565

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java [hunk 1]
```diff
@@ -74,7 +77,7 @@ protected String insertCodeMinings(XtextEditor editor) {
 		ISourceViewer viewer = editor.getInternalSourceViewer();
 
 		String text = editor.getDocument().get();
-		StringBuilder sb = new StringBuilder();
+		StringBuilder sb = new StringBuilder(text);
 
 		List<? extends ICodeMining> codeMinings;
 		try {
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:83]: `Futures.getUnchecked(..)` may simlify this try-catch and the one below.

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java [hunk 2]
```diff
@@ -92,45 +95,21 @@ protected String insertCodeMinings(XtextEditor editor) {
...
-
-			String codeMiningsText = getCodeMiningsText(codeMiningsOnTheSameOffset);
-
-			sb.append(text.substring(currentOffset, codeMiningOffset) + codeMiningsText);
-
-			currentOffset = codeMiningOffset;
-			if (containsLineHeaderCodeMining(codeMiningsOnTheSameOffset)) {
-				sb.append(System.lineSeparator());
-				currentOffset--;
+		Map<Integer, List<ICodeMining>> byPos = codeMinings.stream().collect(Collectors.groupingBy(e -> e.getPosition().getOffset(), LinkedHashMap::new, Collectors.toList()));
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:92]: Note to self: Group the sorted minings -> LinkedHashMap maintains the order.

Comment (szarnekow) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:92]: Maybe more straight forward to use `Multimap.index(codeMinings, e->getPositions().getOffset())` but it's a matter of taste

**Comment by cdietrich** [comment]
@szarnekow updated the PR

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java [hunk 3]
```diff
@@ -74,63 +77,31 @@ protected String insertCodeMinings(XtextEditor editor) {
...
+		for (Entry<Integer, List<ICodeMining>> e : byPos.entrySet()) {
+			int codeMiningOffset = e.getKey();
+			List<ICodeMining> miningsAtOffset = e.getValue();
+			String codeMiningsText = getCodeMiningsText(miningsAtOffset);
+			if (containsLineHeaderCodeMining(miningsAtOffset)) {
+				sb.insert(codeMiningOffset, System.lineSeparator());
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:98]: `System.lineSeparator` -> `((IDocumentExtension4)editor.getDocument()).getDefaultLineDelimiter()`

**Review by szarnekow — APPROVED** [review]

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java (hunk 2 above)
Comment (cdietrich) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:92]: wanted to keep this similat to what eclipse does ;)

**Comment by szarnekow** [comment]
Nice nice! Please go ahead with the merge when Jenkins agrees.

**Pull request merged**